crates/symbolicator/src/services/cacher.rs

```diff
@@ -187,64 +188,58 @@ impl<T: CacheItemRequest> Cacher<T> {
+                    format!("{:?}", path).into(),
+                );
+            },
+            || {
+                let byteview = match self.config.open_cachefile(path)? {
+                    Some(x) => x,
+                    None => return Ok(None),
+                };
+
+                let status = CacheStatus::from_content(&byteview);
+                if status == CacheStatus::Positive && !request.should_load(&byteview) {
+                    log::trace!("Discarding {} at path {:?}", name, path);
+                    metric!(counter(&format!("caches.{}.file.discarded", name)) += 1);
+                    return Ok(None);
+                }
+
+                // This is also reported for "negative cache hits": When we cached the 404 response from a
+                // server as empty file.
+                metric!(counter(&format!("caches.{}.file.hit", name)) += 1);
+                metric!(
+                    time_raw(&format!("caches.{}.file.size", name)) = byteview.len() as u64,
+                    "hit" => "true"
+                );
 
-        let path = path.to_path_buf();
-        log::trace!("Loading {} at path {:?}", name, path);
-        let item = request.load(key.scope.clone(), status, byteview, CachePath::Cached(path));
-        Ok(Some(item))
+                let path = path.to_path_buf();
+                log::trace!("Loading {} at path {:?}", name, path);
+                let item =
+                    request.load(key.scope.clone(), status, byteview, CachePath::Cached(path));
+                Ok(Some(item))
+            },
+        )
     }
 
     /// Compute an item.
     ///
-    /// If the item is in the file system cache, it is returned immediately. Otherwise, it
-    /// is computed using [`T::compute`](CacheItemRequest::compute), and then persisted to
-    /// the cache.
+    /// The item is computed using [`T::compute`](CacheItemRequest::compute), and then persisted to
+    /// the given `cache_path`.
     ///
     /// This method does not take care of ensuring the computation only happens once even
     /// for concurrent requests, see the public [`Cacher::compute_memoized`] for this.
-    async fn compute(&self, request: T, key: CacheKey) -> Result<T::Item, T::Error> {
-        // cache_path is None when caching is disabled.
-        let cache_path = self.get_cache_path(&key);
-        if let Some(ref path) = cache_path {
-            if let Some(item) = self.lookup_cache(&request, &key, path)? {
-                return Ok(item);
-            }
-        }
-
-        let name = self.config.name();
-        let key = request.get_cache_key();
-
-        // A file was not found. If this spikes, it's possible that the filesystem cache
-        // just got pruned.
-        metric!(counter(&format!("caches.{}.file.miss", name)) += 1);
-
-        let temp_file = self.tempfile()?;
+    async fn compute(
+        name: &'static str,
+        tempfile: std::io::Result<NamedTempFile>,
```

Comment:
Your description says you run into lifetime issues when using `&self` and thats why these are here.  But the current arguments are a bit confusing and I had to look at both versions to understand them.  I wonder if passing `self` is another reasonable stopgap here?  It does mean you have to clone all of self, but `Cacher::current_computations` is already an `Arc` and `Cache` is... still clonable but not super cheap.

Reply:
I think the problem basically was that somewhere, we require these futures to be `Send + 'static`, which meant that also `&self` needed to be `'static`, which essentially means you need to take it by value. I remember we had that problem in other places as well, but here I opted to just explicitly call the async fn (create the future) with just the types that are needed inside. The `tempfile` is a bit unfortunate though :-(

Reply:
Oh, so the old code already cloned. Oh well, I changed it now so that it continues cloning. It looks like the cleanest solution right now.